### PR TITLE
community: Add Bedrock chat_model async methods #11668

### PR DESCRIPTION
**Pull request by villekr**
- **Description:** Add async methods for Bedrock chat_model and complete the earlier work in https://github.com/langchain-ai/langchain/pull/12477. Attemps to reduce some code duplication. 
- **Issue:** https://github.com/langchain-ai/langchain-aws/issues/171
- **Dependencies:** 
- **Twitter handle:** @villekr

### CODE REVIEW

**Comment by villekr** [comment]
I'm using the following test code (modified from https://github.com/langchain-ai/langchain/pull/12477#issuecomment-1898612350)

If you run it with the current langchain version (0.1.5 in pypi) the BedrockChat variant will not progress concurrently. But with this PR version it does.

```
import asyncio

from langchain_community.llms import Bedrock
from langchain_community.chat_models import BedrockChat
from langchain.schema.output_parser import StrOutputParser
from langchain.prompts import ChatPromptTemplate

llm_model = Bedrock(model_id="anthropic.claude-v2:1")
chat_model = BedrockChat(model_id="anthropic.claude-v2:1")
prompt = ChatPromptTemplate.from_template("Tell me a story about {topic}.")
parser = StrOutputParser()


async def main():
    await asyncio.gather(
        primary_task(llm_model),
        secondary_task("with Bedrock"),
    )

    await asyncio.gather(
        primary_task(chat_model),
        secondary_task("with BedrockChat"),
    )


async def primary_task(model):
    chain = prompt | model | parser
    async for chunk in chain.astream({"topic": "bears"}):
        print(chunk)


async def secondary_task(name: str):
    print(f"- secondary_task {name} start")
    for i in range(10):
        await asyncio.sleep(1)
        print(f"-- secondary_task {name} increment {i}")


if __name__ == "__main__":
    asyncio.run(main())
```

**Comment by villekr** [comment]
@baskaryan, @eyurtsev, @hwchase17 Just let me know if there's anything that keeps this PR awaiting for review & merge.

**Comment by baskaryan** [comment]
cc @3coins

**Comment by yw791772** [comment]
Hi, is there any updates here, as of right now, bedrock streaming is not working as expected, it's using the invoke model method that aws provided, not the InvokeModelWithResponseStream

**Review by jaswanth8888 — APPROVED** [review]

**Comment by villekr** [comment]
Looks like there are some workflows that require approvals before merge is possible. @jaswanth8888 how to proceed?

**Comment by villekr** [comment]
@baskaryan, @eyurtsev, @hwchase17, @jaswanth8888 please advice how to get this change merged. Workflows are awaiting approval (and there's no way to merge) although the PR is approved.

**Comment by villekr** [comment]
Sorry to repeat. Please take care of merging as this has been ready for quite a long time and is approved. @baskaryan, @eyurtsev, @hwchase17, @jaswanth8888.

**Comment by villekr** [comment]
It's almost *a month* since this PR was opened. I know we all are busy but keeping ready and approved improvements pending such a long time is just frustrating. @baskaryan, @eyurtsev, @hwchase17, @3coins, @jaswanth8888.

File: libs/community/langchain_community/llms/bedrock.py
Comment (baskaryan) [inline libs/community/langchain_community/llms/bedrock.py:502]: is this an async method? as far as i can tell it's not

is there any other difference between this and _prepare_input_and_invoke?

Comment (villekr) [reply libs/community/langchain_community/llms/bedrock.py:502]: Excellent catch! I was basically interested in async streaming and while I added that _aprepare_input_and_invoke I missed to actually test that. 

I'll make a fix for that.

**Comment by villekr** [comment]
Here's the updated code I used to test this PR.

```
import asyncio

from langchain_community.llms import Bedrock
from langchain_community.chat_models import BedrockChat
from langchain.schema.output_parser import StrOutputParser
from langchain.prompts import ChatPromptTemplate

llm_model = Bedrock(model_id="anthropic.claude-v2:1")
chat_model = BedrockChat(model_id="anthropic.claude-v2:1")
prompt = ChatPromptTemplate.from_template("Tell me a story about {topic}.")
parser = StrOutputParser()


async def main():
    # langchain 0.1.11: raises ValueError: Streaming must be set to True for async operations.
    # PR: works as expected
    await asyncio.gather(
        primary_task_ainvoke(llm_model),
        secondary_task("Async invoke with Bedrock", 3),
    )

    # langchain 0.1.11: works as expected
    # PR: works as expected
    await asyncio.gather(
        primary_task_ainvoke(chat_model),
        secondary_task("Async invoke with BedrockChat", 3),
    )

    # langchain 0.1.11: works as expected
    # PR: works as expected
    await asyncio.gather(
        primary_task_astream(llm_model),
        secondary_task("Async stream with Bedrock"),
    )

    # langchain 0.1.11: produced output but is not async
    # PR: works as expected
    await asyncio.gather(
        primary_task_astream(chat_model),
        secondary_task("Async stream with BedrockChat"),
    )


async def primary_task_ainvoke(model):
    chain = prompt | model | parser
    response = await chain.ainvoke({"topic": "bears"})
    print(response)


async def primary_task_astream(model):
    chain = prompt | model | parser
    async for chunk in chain.astream({"topic": "bears"}):
        pass
        print(chunk)


async def secondary_task(name: str, delay: int = 1):
    print(f"- secondary_task {name} start")
    for i in range(10):
        await asyncio.sleep(delay)
        print(f"-- secondary_task {name} increment {i}")


if __name__ == "__main__":
    asyncio.run(main())

```

**Comment by villekr** [comment]
Here's the summary of the subset of Bedrock operations' status based on my testin:

| Operation            | Langchain 0.1.11 | This PR |
|----------------------|------------------|---------|
| llm ainvoke          | ❌ NOK(1)           | ✅ OK    |
| chat-model ainvoke   | ✅ OK            | ✅ OK    |
| llm astream          | ✅ OK            | ✅ OK    |
| chat-model astream   | ❌ NOK (2)           | ✅ OK    |

* 1 - raises ValueError: Streaming must be set to True for async operations.
* 2 - produced output but is not async

I modified the Bedrock._call function, which was mentioned [here](https://github.com/langchain-ai/langchain/pull/12477#discussion_r1459585389).

There would still be some room for cleaning up the code and remove duplicate but I'd leave that later.

**Comment by villekr** [comment]
Latest update contains:
* Resolved merge conflicts. There seems to have some recent changes for bedrock in master-branch.
* Refactored to remove some code duplication

Updated also my test code snippet to include calls for sync-methods. Here are the results with PR vs. master-branch/published version.

| Operation                   | Langchain 0.1.11  | This PR |
|                                    | master-branch    |               |
|----------------------|------------------|---------|
| llm invoke                   | ✅ OK           | ✅ OK    |
| chat-model invoke     | ✅ OK           | ✅ OK    |
| llm stream                  | ✅ OK           | ✅ OK    |
| chat-model stream    | ✅ OK           | ✅ OK    |
| llm ainvoke                 | ❌ NOK(1)    | ✅ OK    |
| chat-model ainvoke   | ✅ OK           | ✅ OK    |
| llm astream                | ✅ OK           | ✅ OK    |
| chat-model astream  | ❌ NOK (2)  | ✅ OK     |

* 1 - raises ValueError: Streaming must be set to True for async operations.
* 2 - produced output but is not async

```
import asyncio

from langchain.prompts import ChatPromptTemplate
from langchain.schema.output_parser import StrOutputParser
from langchain_community.chat_models import BedrockChat
from langchain_community.llms import Bedrock

llm_model = Bedrock(model_id="anthropic.claude-v2:1")
chat_model = BedrockChat(model_id="anthropic.claude-v2:1")
prompt = ChatPromptTemplate.from_template("Tell me a story about {topic}.")
parser = StrOutputParser()


async def main():
    # Sync operations

    primary_task_invoke(llm_model, "Sync invoke with llm_model")
    primary_task_invoke(chat_model, "Sync invoke with chat_model")
    primary_task_stream(llm_model, "Sync stream with llm_model")
    primary_task_stream(chat_model, "Sync stream with chat_model")

    # Async operations

    # langchain 0.1.11: raises ValueError: Streaming must be set to True for async operations.  # noqa: B950
    # langchain master: (above)
    # PR: works as expected
    try:
        await asyncio.gather(
            primary_task_ainvoke(llm_model, "Async invoke with llm_model"),
            secondary_task("Async invoke with llm_model", 3),
        )
    except Exception as e:
        print(e)

    # langchain 0.1.11: works as expected
    # langchain master: (above)
    # PR: works as expected
    await asyncio.gather(
        primary_task_ainvoke(chat_model, "Async invoke with chat_model"),
        secondary_task("Async invoke with chat_model", 3),
    )

    # langchain 0.1.11: works as expected
    # langchain master: (above)
    # PR: works as expected
    await asyncio.gather(
        primary_task_astream(llm_model, "Async stream with llm_model"),
        secondary_task("Async stream with llm_model", 3),
    )

    # langchain 0.1.11: produced output but is not async
    # langchain master: (above)
    # PR: works as expected
    await asyncio.gather(
        primary_task_astream(chat_model, "Async stream with chat_model"),
        secondary_task("Async stream with chat_model", 3),
    )


def primary_task_invoke(model, name: str):
    print(f"* primary_task_invoke {name} start")
    chain = prompt | model | parser
    response = chain.invoke({"topic": "bears"})
    print(response)


def primary_task_stream(model, name: str):
    print(f"* primary_task_stream {name} start")
    chain = prompt | model | parser
    for chunk in chain.stream({"topic": "bears"}):
        print(chunk)


async def primary_task_ainvoke(model, name: str):
    print(f"* primary_task_ainvoke {name} start")
    chain = prompt | model | parser
    response = await chain.ainvoke({"topic": "bears"})
    print(response)


async def primary_task_astream(model, name: str):
    print(f"* primary_task_astream {name} start")
    chain = prompt | model | parser
    async for chunk in chain.astream({"topic": "bears"}):
        pass
        print(chunk)


async def secondary_task(name: str, delay: int = 1):
    print(f"- secondary_task {name} start")
    for i in range(10):
        await asyncio.sleep(delay)
        print(f"-- secondary_task {name} increment {i}")


if __name__ == "__main__":
    asyncio.run(main())

```

**Comment by villekr** [comment]
Fixed format and lint issues as well as merge conflicts with master. Based on my testing also works ok.

**Comment by villekr** [comment]
@baskaryan can you help review and merge this finally? It's quite important to have async chat_model methods available.

**Comment by villekr** [comment]
I fixed merge conflicts once again.

Code quality checks are ok as well as my own tests from Bedrock llm and chat-model sync and async requests. Please proceed for review and hopefully this gets merged soon.

**Comment by chunyeah** [comment]
Hi Thanks for this PR. 👍
When I use this PR code, still got some error multiple requests below:

WARNING [langchain_core.callbacks.manager._ahandle_event_for_handler:381] Error in CharacterAsyncHandler.on_llm_new_token callback: ClosedResourceError()

**Comment by villekr** [comment]
Pleasr provide example code where this warning is visible and reproducible.

**Comment by chunyeah** [comment]
> Pleasr provide example code where this warning is visible and reproducible.
> 
> Do you get that warning only with this PR code and not with master-branch version?

Here's your translation:

Thank you for your reply. I am not using the bedrock code on master, as the online environment does not support asynchronous operations and cannot function. After discovering the warning mentioned, I tried to adjust the synchronous to asynchronous code as follows:

```python
with ThreadPoolExecutor() as executor:
    response = await loop.run_in_executor(
        executor,
        lambda: self.client.invoke_model_with_response_stream(
            **request_options
        ),
    )
```

Even so, when the code runs in the online environment, it causes the pod resources to increase very rapidly. I am not sure if it is the synchronous-to-asynchronous conversion causing this. Other than this change, all other code is from the code in this PR.

Thank you for your work.

**Comment by chunyeah** [comment]
> I did a bit of testing and memory profiling comparing Bedrock and OpenAI async streaming. With my own test code I'm not able to reproduce the issue you refer. Also I don't see any specific issues with memory usage, which might cause increased resource usage you mention. What's the online environment you refer to?
> 
> I don't know how to approach your problem without code I can use to replicate the issue.
> 
> Here's the code I used in my test.
> 
> ```
> import asyncio
> from typing import Any
> 
> from langchain.callbacks.base import AsyncCallbackHandler
> from langchain.prompts import ChatPromptTemplate
> from langchain.schema.output import LLMResult
> from langchain.schema.output_parser import StrOutputParser
> from langchain_community.chat_models import BedrockChat
> from langchain_community.llms import Bedrock
> from langchain_openai import ChatOpenAI, OpenAI
> from memory_profiler import profile
> 
> 
> class AsyncHandler(AsyncCallbackHandler):
>     async def on_llm_new_token(self, token: str, *args, **kwargs: Any) -> None:
>         # print(token)
>         pass
> 
>     async def on_llm_end(self, response: LLMResult, *args, **kwargs: Any) -> None:
>         # print(response)
>         pass
> 
>     async def on_llm_error(self, *args, **kwargs: Any) -> None:
>         pass
> 
> 
> async_callback_handler = AsyncHandler()
> 
> bedrock_chat_model = BedrockChat(
>     model_id="anthropic.claude-v2:1", streaming=True, callbacks=[async_callback_handler]
> )
> openai_chat_model = ChatOpenAI(
>     model_name="gpt-4", streaming=True, callbacks=[async_callback_handler]
> )
> 
> prompt = ChatPromptTemplate.from_template("Tell me a story about {topic}.")
> parser = StrOutputParser()
> 
> 
> async def main():
>     await astream(openai_chat_model, "OpenAI Chat")
>     await astream(bedrock_chat_model, "Bedrock Chat")
> 
> 
> @profile(precision=4)
> async def astream(model, name: str):
>     print(f"{name} start")
>     chain = prompt | model | parser
>     async for _ in chain.astream({"topic": "bears"}):
>         pass
> 
> 
> if __name__ == "__main__":
>     asyncio.run(main())
> ```
> 
> ```
> > poetry run python langchain_11668/langchain_11668_memory_profile.py
> OpenAI Chat start
> Filename: /.../langchain_11668_memory_profile.py
> 
> Line #    Mem usage    Increment  Occurrences   Line Contents
> =============================================================
>     45 100.8125 MiB 100.8125 MiB           1   @profile(precision=4)
>     46                                         async def astream(model, name: str):
>     47 100.8125 MiB   0.0000 MiB           1       print(f"{name} start")
>     48 100.8125 MiB   0.0000 MiB           1       chain = prompt | model | parser
>     49 102.9531 MiB -32699.1719 MiB        1210       async for _ in chain.astream({"topic": "bears"}):
>     50 102.9531 MiB -16322.1719 MiB         604           pass
> 
> 
> Bedrock Chat start
> Filename: /...langchain_11668_memory_profile.py
> 
> Line #    Mem usage    Increment  Occurrences   Line Contents
> =============================================================
>     45  46.9062 MiB  46.9062 MiB           1   @profile(precision=4)
>     46                                         async def astream(model, name: str):
>     47  46.9062 MiB   0.0000 MiB           1       print(f"{name} start")
>     48  47.3125 MiB   0.4062 MiB           1       chain = prompt | model | parser
>     49  66.9844 MiB -826.5469 MiB         346       async for _ in chain.astream({"topic": "bears"}):
>     50  66.9844 MiB -419.7500 MiB         172           pass
> ```

Thanks your reply. I will try to streamline the code and post it, thank you.

**Comment by derekhsu** [comment]
There is no `_aprepare_input_and_invoke_stream` method in the class `ChatBedrock` :(

File: libs/community/langchain_community/llms/bedrock.py
Comment (derekhsu) [reply libs/community/langchain_community/llms/bedrock.py]: I don't see `_aprepare_input_and_invoke` in the committed files, didn't miss anything?

**Comment by yw791772** [comment]
any update on this one?

**Comment by ccurme** [comment]
@3coins would you mind taking a look at this? Possible the issue exists in langchain-aws as well.

**Comment by efriis** [comment]
Hey team! This touches the deprecated community bedrock integration, which has been replaced by the `langchain-aws` package (co-managed with the bedrock team)

This only runs the sync methods in an executor, which the BaseChatModel should be doing anyways, and if that's not the case, please include reproduction code in the issue

I'm going to move the issue to their repo for review there and reference this PR.

**Pull request closed without merging**
